### PR TITLE
Deploy multiple modules (unique instances)

### PR DESCRIPTION
**Pull request by isaacroldan**
### WHY are these changes introduced?

Fixes 
- https://github.com/Shopify/develop-app-management/issues/1720
- https://github.com/Shopify/develop-app-management/issues/1724 
- https://github.com/Shopify/develop-app-management/issues/1776

See https://github.com/Shopify/develop-app-management/issues/1720 for more details and 🎩  instructions

### Measuring impact

How do we know this change was effective? Please choose one:

- [ ] n/a - this doesn't need measurement, e.g. a linting rule or a bug-fix
- [ ] Existing analytics will cater for this addition
- [ ] PR includes analytics changes to measure impact

### Checklist

- [ ] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [ ] I've considered possible [documentation](https://shopify.dev) changes
- [ ] I've made sure that any changes to `dev` or `deploy` have been reflected in the [internal flowchart](https://www.figma.com/file/7vqUp50u6dm48Zfb4JRRn8/CLI3-Internals).

### CODE REVIEW

File: packages/app/src/cli/services/context/identifiers-extensions.ts
Comment (gracejychang) [inline packages/app/src/cli/services/context/identifiers-extensions.ts:154]: we don't need to transform since `config` is just the subscription now

File: packages/app/src/cli/models/app/app.ts
Comment (isaacroldan) [inline packages/app/src/cli/models/app/app.ts:335]: is this method `filterDeclarativeWebhooksConfig` actually necessary?
The specification is filtered in the backend, if the CLI doesn't receive the spec it won't load any subscription. So there is no need to filter them here.

But do we need to remove them from the `webhooks` and `privacy-compliance-webhooks` too?

File: packages/app/src/cli/services/app/select-app.ts
Comment (isaacroldan) [inline packages/app/src/cli/services/app/select-app.ts:54]: can't we use `spec.uidStrategy === 'single'` here? That will include the `webhooks` specs which already has the transform code for this and should be enough no?

Comment (isaacroldan) [reply packages/app/src/cli/services/app/select-app.ts:54]: I tested by using `=== 'single'` here and the link command works fine, should we change it then?

File: packages/app/src/cli/models/extensions/specifications/app_config_webhook_subscription.ts
Comment (isaacroldan) [reply packages/app/src/cli/models/extensions/specifications/app_config_webhook_subscription.ts]: Fixed

File: packages/app/src/cli/services/context/identifiers-extensions.ts
Comment (alexanderMontague) [reply packages/app/src/cli/services/context/identifiers-extensions.ts:159]: I also think this is another spot where `filters` will need to be added. I wonder if there is a better abstraction for uniqueness properties. Not a blocker

File: packages/app/src/cli/services/context/identifiers-extensions.ts
Comment (alexanderMontague) [inline packages/app/src/cli/services/context/identifiers-extensions.ts:206]: I don't think this needs to be awaited/async either

File: packages/app/src/cli/services/context/identifiers-extensions.ts
Comment (alexanderMontague) [inline packages/app/src/cli/services/context/identifiers-extensions.ts:180]: This is also something very webhooks specific, could this also be abstracted away to the generic interface? This logic will also move to the server eventually though right? Not a blocker

**Review by alexanderMontague — COMMENTED** [review]
I think this is great, well done! Would love to get this shipped asap so most of my feedback is comments/improvements/questions and mostly non blocking. Will approve once addressed!

File: packages/app/src/cli/models/app/loader.ts
Comment (isaacroldan) [reply packages/app/src/cli/models/app/loader.ts:565]: it could be in the original webhooks schema, and I guess that's something we don't want to change now

File: packages/app/src/cli/models/extensions/extension-instance.ts
Comment (isaacroldan) [reply packages/app/src/cli/models/extensions/extension-instance.ts:149]: We didn't want to change all the calls to this, but we should change it yes

File: packages/app/src/cli/models/extensions/extension-instance.ts
Comment (isaacroldan) [reply packages/app/src/cli/models/extensions/extension-instance.ts:355]: Eventually yes, we will need to use filters too as part of the unique identifier

File: packages/app/src/cli/models/extensions/extension-instance.ts
Comment (isaacroldan) [reply packages/app/src/cli/models/extensions/extension-instance.ts:356]: I mean, I don't expect an app to have 1000s of webhooks no? It should be safe, I hope :P

File: packages/app/src/cli/services/context/identifiers-extensions.test.ts
Comment (isaacroldan) [reply packages/app/src/cli/services/context/identifiers-extensions.test.ts:1053]: I renamed it yes, i'll fix ti

File: packages/app/src/cli/services/context/identifiers-extensions.ts
Comment (isaacroldan) [reply packages/app/src/cli/services/context/identifiers-extensions.ts:180]: The idea was to not modify the generic interface at all and keep all the webhooks stuff isolated so that we can easily remove it in the future.

File: packages/app/src/cli/models/app/loader.ts
Comment (alexanderMontague) [reply packages/app/src/cli/models/app/loader.ts:582]: ah correct, because of the string literal type. disregard 👍

File: packages/app/src/cli/services/context/identifiers-extensions.ts
Comment (alexanderMontague) [reply packages/app/src/cli/services/context/identifiers-extensions.ts:180]: makes sense 👍

File: packages/app/src/cli/models/app/app.ts
Comment (gracejychang) [reply packages/app/src/cli/models/app/app.ts]: I think removing this will cause `webhook_subscriptions` registrations to be created in the database even if the beta is off - I did a 🎩 to confirm and saw the `webhook_subscriptions` registration in the db & in the network call after I `deploy` with the beta changes removed. I think it doesn't show up on the Versions page because there is also an additional beta check on the frontend

on Versions Page:
Screenshot 2024-05-09 at 2 10 32 PM

Network call, the `handle` with the hashstring is the webhook subscription module
Screenshot 2024-05-09 at 2 11 26 PM

Database 
Screenshot 2024-05-09 at 2 12 50 PM

File: packages/app/src/cli/services/context/identifiers-extensions.test.ts
Comment (gracejychang) [reply packages/app/src/cli/services/context/identifiers-extensions.test.ts:1111]: I believe so, since in `testDeveloperPlatformClient`, `createExtension` is a function that resolves a Promise:

```
createExtension: (_input: ExtensionCreateVariables) => Promise.resolve(extensionCreateResponse),
```

Comment (alexanderMontague) [reply packages/app/src/cli/services/context/identifiers-extensions.test.ts:1111]: shouldn't it just infer that `createExtensionResult` returns a Promise? weird

File: packages/app/src/cli/models/app/loader.ts
Comment (gracejychang) [reply packages/app/src/cli/models/app/loader.ts:576]: to be honest, there are a lot of promises in CLI and I also get confused by them 😅  @isaacroldan do you have some insight here?

File: packages/app/src/cli/services/context/identifiers-extensions.test.ts
Comment (gracejychang) [reply packages/app/src/cli/services/context/identifiers-extensions.test.ts]: renamed in this [commit](https://github.com/Shopify/cli/pull/3846/commits/e05a325310f6d2fecb9f7e0361d554bc83bbe2bb) 😄

File: packages/app/src/cli/models/app/loader.ts
Comment (alexanderMontague) [reply packages/app/src/cli/models/app/loader.ts:576]: not a big deal, its fine to get this in, but ideally this gets cleaned up sometime :)

File: packages/app/src/cli/models/app/loader.ts
Comment (gracejychang) [reply packages/app/src/cli/models/app/loader.ts:551]: hmmm it looks like this is only being used once when we are calling `createExtensionInstance`, so I don't think it's necessary

Comment (gracejychang) [reply packages/app/src/cli/models/app/loader.ts]: I've removed it and just called `appWebhookSubscriptionSpec` directly where we were using `specification` before

File: packages/app/src/cli/models/extensions/extension-instance.ts
Comment (gracejychang) [reply packages/app/src/cli/models/extensions/extension-instance.ts]: Removed it in this [commit](https://github.com/Shopify/cli/pull/3846/commits/e797b085adce624beefc051cf7d14bc38c0db7a0)

File: packages/app/src/cli/models/app/loader.ts
Comment (isaacroldan) [reply packages/app/src/cli/models/app/loader.ts]: I cleaned up the promises a bit

File: packages/app/src/cli/services/context/identifiers-extensions.ts
Comment (karenxie) [inline packages/app/src/cli/services/context/identifiers-extensions.ts:156]: should we remove the `draftVersion`? since `dev `won't work with webhook subscriptions

**Review by karenxie — APPROVED** [review]
Code LGTM! We should get a CLI expert approval too. Also, given that there have been a bunch of little changes since we last tested things, I'll give it another whirl this afternoon

File: packages/app/src/cli/models/extensions/extension-instance.ts
Comment (dpeacock) [reply packages/app/src/cli/models/extensions/extension-instance.ts]: For reference the [filters PR](https://github.com/Shopify/cli/pull/3862/) was merged through this morning.  The [uniqueness check is here](https://github.com/Shopify/cli/blob/main/packages/app/src/cli/models/extensions/specifications/validation/app_config_webhook.ts#L53) currently.

File: packages/app/src/cli/models/extensions/specifications/app_config_webhook_subscription.ts
Comment (gracejychang) [reply packages/app/src/cli/models/extensions/specifications/app_config_webhook_subscription.ts:24]: it seems like `filter` can be optional, so are you ok with adding this as a fast follow? Reason being, we will change have to change some other parts of the code to include `filter` (ie. matching logic) too, and there is already a lot to 🎩 and keep track of in this PR

File: packages/app/src/cli/models/app/app.test-data.ts
Comment (isaacroldan) [reply packages/app/src/cli/models/app/app.test-data.ts:267]: I just duplicated the function from above 😅 
There is no need to load all extensions. We originally had this because you couldn't import a single specification (they weren't "exported") so you had to load all.
But that has changed and is now possible to import specific ones

**Pull request merged**
